### ISSUE TITLE
Full call stack lost with Async services

### ISSUE THREAD

**Issue by riesvriend** [ServiceStack/Issues#387]
I am writing my first Task based async servicestack service. It's goal is to deploy code to AWS using the AWS APIs. All that code is async as well. The problem is that when an exception occurs, by default, the client only receives the inner-most exception message and the inner-most call stack. All call context that is helpful to pinpoint and debug the issue is lost.

I found a workaround, which is to catch the exception at the topmost (service) method, and re-throwing it with a copy of all inner exception messages and with the call stack, which is actually including the call stack of the awaiting method chain.

Can this behavior be enabled in servicestack by default, somehow?

```c#
// Problem
public async Task<DescribeDeploymentsResponse> Get(DescribeDeploymentsRequest request)
{
    var fbo1Manager = new Fbo1ManangerService();
    return await fbo1Manager.DescribeDeployments(request);
}
```
outputs: [problem.txt](https://github.com/ServiceStack/Issues/files/74847/problem.txt)

```c#
// Workaround
public async Task<DescribeDeploymentsResponse> Get(DescribeDeploymentsRequest request)
{
  try
   {
     var fbo1Manager = new Fbo1ManangerService();
     return await fbo1Manager.DescribeDeployments(request);
   }
  catch (Exception ex)
  {
    var s = ex.StackTrace;
    throw new Exception(ex.AllMessages() + " Stack: " + s, ex);
  }
}
```
Outputs all calling methods (towards the end of the file): [workaround-exception.txt](https://github.com/ServiceStack/Issues/files/74846/workaround-exception.txt)

**Comment by mythz** [comment]
If you don't want the inner exception you specify it with:

```csharp
SetConfig(new HostConfig {
    ReturnsInnerException = false
});
```

Let me know if that resolves your issue?

**Comment by riesvriend** [comment]
Hi Demis,

Yes that fixes it, thanks for the help. 

Looking at the source it seems that by default the outer-most exception is always discarded. 

https://github.com/ServiceStack/ServiceStack/blob/master/src/ServiceStack/HostConfig.cs (line 85)
https://github.com/ServiceStack/ServiceStack/blob/master/src/ServiceStack/DtoUtils.cs (line 143)

It comes across as rather risky to me. Would it not be better to have (configurable) a list of exception types to discard or keep (as in the `!(ex is IHttpError)` in line 144)?

Wishing you and your team a great 2016.

Ries

**Comment by mythz** [comment]
> Would it not be better to have a (configurable) list of exception types to discard or keep, as in the !(ex is IHttpError) in line 144?

That would be too opinionated to have baked in the fx, but I've made it so the response exception behavior can be customized in this commit: https://github.com/ServiceStack/ServiceStack/commit/db607a29cd84a810f097fa726b46509e42603d8where you can now override `ResolveResponseException()` in your AppHost to override the existing behavior.

This change is available from v4.0.51 that's now [available on MyGet](https://github.com/ServiceStack/ServiceStack/wiki/MyGet).
